### DISCUSSION TITLE
Row-type polymorphism in the context of Cardano scripts

### DISCUSSION THREAD

**Discussion by klntsky** [mlabs-haskell/purus#2] [General]
Regarding row type polymorphism (RTP for short from now on), which is the only non-trivial language feature that we have to implement, we have to consider two questions:

I. What value does it bring to the PS ecosystem and how is it relevant for the on-chain scripts domain
II. What costs are we willing to pay for RTP?

**Comment by klntsky** [comment]
> I. What value does it bring to the PS ecosystem and how is it relevant for the on-chain scripts domain

As far as I'm aware of the ecosystem, row type polymorphism is mainly used to achieve the following goals:

 1. Operations with records bypassing built-in syntactic constructs: e.g. [updating record fields using Proxy](https://pursuit.purescript.org/packages/purescript-record/4.0.0/docs/Record) - this approach is of course more composable than record update syntax. Also, lenses.
 2. Defining data-generic encoders/decoders for record types (most notably, [argonaut](https://pursuit.purescript.org/packages/purescript-argonaut-codecs/9.1.0/docs/Data.Argonaut.Decode.Class#t:GDecodeJson) for JSON)
 3. FFI stuff for simpler JS interop, e.g. [untagged-union](https://github.com/rowtype-yoga/purescript-untagged-union/blob/ed8262a966e15e751322c327e2759a9b9c0ef3f3/src/Untagged/Castable.purs#L26)
 4. Custom data machinery, like type-safe [variants](https://pursuit.purescript.org/packages/purescript-variant/8.0.0/docs/Data.Variant) - this is a clever hack that is arguably not widely used (no core libraries depend on it)

**Reply by klarkc** [reply]
I believe RTP is highly valuable when decoding/encoding JSON, which is very common in web development - a context where PureScript is likely to be used. I suggest not giving up on RTP, or at least making it a low priority / future effort.

**Comment by klntsky** [comment]
>  II. What costs are we willing to pay for RTP?

I think that if we fail to find a way to implement RTP as a zero-runtime-cost abstraction, we should give up on it, because, as we know from Haskell experience, using a FP language with no RTP is *just fine*. And we are not even building a general purpose language - just a glorified DSL.

**Reply by klntsky** [reply]
Let's examine the runtime costs of the approaches/libs involing RTP I listed [here](https://github.com/mlabs-haskell/purus/discussions/2#discussioncomment-7422966) with the existing JS backend. My argument is: if the PS folks didn't figure out a zero-runtime-cost approach for JS, *then* it is *unproductive* for us to try doing that for UPLC, given the limited budget. A few years ago even Proxy was not zero-cost, but we need to figure out the current state of affairs.

**Reply by klarkc** [reply]
AFAIK, records in PureScript translate to plain runtime JS objects, enabling unique optimizations at runtime, beyond just compile time.

**Comment by paluh** [comment]
In practice having polymorphic types is also useful to share type definitions like:

```
type Base r = { a :: A | r }
type R1 = Base (b :: B)
type R2 = Base (c :: C)
```

I wonder if it is feasible to keep row RTP in the type system but require full manual specialization of value types.

> I think that if we fail to find a way to implement RTP as a zero-runtime-cost abstraction, we should give up on it (..)

I fully agree with this if I can share my user opinion. Maybe even smaller language (better error messages?) would be easier to use in this context.
